chia/wallet/wallet_node.py

```diff
@@ -466,41 +466,40 @@ async def _messages_to_resend(self) -> List[Tuple[Message, Set[bytes32]]]:
-                                continue
-                            else:
-                                self.log.info(
-                                    f"disconnected from peer {peer_id}, state will retry with {peer.peer_node_id}"
-                                )
+                if self.wallet_state_manager is None:
+                    continue
+                states_to_retry = await self.wallet_state_manager.retry_store.get_all_states_to_retry()
+                for state, peer_id, fork_height in states_to_retry:
+                    matching_peer = tuple(
+                        p for p in self.server.get_connections(NodeType.FULL_NODE) if p.peer_node_id == peer_id
+                    )
+                    if len(matching_peer) == 0:
+                        peer = self.get_full_node_peer()
+                        if peer is None:
+                            self.log.info(f"disconnected from all peers, cannot retry state: {state}")
+                            continue
                         else:
-                            peer = matching_peer[0]
-                        async with self.wallet_state_manager.db_wrapper.writer():
-                            self.log.info(f"retrying coin_state: {state}")
-                            try:
-                                await self.wallet_state_manager.new_coin_state(
-                                    [state], peer, None if fork_height == 0 else fork_height
-                                )
-                            except Exception as e:
-                                self.log.exception(f"Exception while adding states.. : {e}")
-                            else:
-                                await self.wallet_state_manager.blockchain.clean_block_records()
+                            self.log.info(
+                                f"disconnected from peer {peer_id}, state will retry with {peer.peer_node_id}"
+                            )
+                    else:
+                        peer = matching_peer[0]
+                    async with self.wallet_state_manager.db_wrapper.writer():
+                        self.log.info(f"retrying coin_state: {state}")
+                        try:
+                            await self.wallet_state_manager.new_coin_state(
+                                [state], peer, None if fork_height == 0 else fork_height
+                            )
+                        except Exception as e:
+                            self.log.exception(f"Exception while adding states.. : {e}")
+                        else:
+                            await self.wallet_state_manager.blockchain.clean_block_records()
             except asyncio.CancelledError:
                 self.log.info("Retry task cancelled, exiting.")
                 raise
+            finally:
+                if not self._shut_down:
+                    await asyncio.sleep(self.coin_state_retry_seconds)
```

Comment:
this block looks weird to me. I would expect this `sleep()` as the last statement inside the try-catch block. Putting it in a finally block means we'll sleep when any exception is raised, before re-raising. Even `CancelledError` will have to sleep. The only exception is if `self._shut_down` is `True`.

Reply:
> this block looks weird to me. I would expect this `sleep()` as the last statement inside the try-catch block. Putting it in a finally block means we'll sleep when any exception is raised, before re-raising. Even `CancelledError` will have to sleep. The only exception is if `self._shut_down` is `True`.

Hm true, it will also sleep for other exceptions but i think in canceled case we will skip it because it only is canceled if `self._shut_down` is `True`, at least thats why i added it :) Initially i checked if `self._retry_failed_states_task.canceled()` is `True` to skip the sleep. But anyway, your right something needs to be changed here!